docs/design/constraints.md

```diff
@@ -0,0 +1,177 @@
+    }
+```
+When creating the constraint, it is assumed that constraint will be initialized and its context will be evaluated.
+Example: if constraint needs to know the project MSBuild properties it is assumed that project will be identified and MSBuild properties will be evaluated when the constraint is being created prior to first evaluation.
+Template constraints initialization will be started in parallel once `TemplateConstraintManager` is created. The methods that work with constraint will ensure task completion before evaluation.
+
+`ITemplateConstraint` definition:
+```csharp
+    public interface ITemplateConstraint
+    {
+        /// <summary>
+        /// Unique constraint type. Should match the definition in `template.json`.
+        /// </summary>
+        string Type { get; }
+
+        /// <summary>
+        /// The user friendly constraint name, that can be used in UI.
+        /// </summary>
+        string DisplayName { get; }
+
+        /// <summary>
+        /// Evaluates <see cref="ITemplateInfo"/> and returns <see cref="TemplateConstraintResult"/> containing result.
+        /// </summary>
+        TemplateConstraintResult Evaluate(string args);
+    }
+
+```
+where `args` contains arguments for constraint read from `template.json`. `Type` matches `type` from constraint configuration (case-sensitive).
+Note: the parameter type may be extended if needed. 
+`Evaluate` method should have optimal performance and operate on the context already evaluated when the constraint was created. It should not launch any time-consuming operation.
+`TemplateConstraintResult` contains the following properties:
+- `Allowed`: boolean - the template is allowed in the given context.
+- `LocalizedErrorMessage`: string - localized error message in case the template is restricted (when `Allowed == false`).
+- `CallToAction`: string - localized message that indicate the action to be taken so that the constraint is fulfilled (optional).
+
+Examples for constraints above:
+- template being run on Windows
+    - `Allowed`: `false`
+    - `LocalizedErrorMessage`: `The template can only be run on the following operating systems: Linux, MacOS.`
+    - `CallToAction`: `null`
+- Xamarin constraint is not met
+    - `Allowed`: `false`
+    - `LocalizedErrorMessage`: `The template requires Xamarin component [1.0.0.0 - 5.0.0.0].`
+    - `CallToAction`: `Install Xamarin component [1.0.0.0 - 5.0.0.0] using Visual Studio installer.`
+
+## API to work with constraints
+
+### `TemplateConstraintManager` class
+
+Class works similar to `TemplatePackageManager`. Once created, it starts initialization of template constraints in background. The host has to create it passing `IEngineEnvironmentSettings`. To reinitialize constraints, new instance of class should be created. It is assumed that the host creates it once, and uses it until disposed.
+
+So far, 2 methods are planned:
+
+`Task<TemplateConstraintResult> EvaluateConstraintAsync(string type, string args)`
```

Comment:
Nit: Args will always contain a JSON, right? Maybe
```suggestion
`Task<TemplateConstraintResult> EvaluateConstraintAsync(string type, string jsonArgs)`
```

Reply:
In theory, it can be any type. Depends on constraint implementation.